### DISCUSSION TITLE
这个 which 依赖 是阻止静态编译的唯一原因

### DISCUSSION THREAD

**Discussion by PikuZheng** [mokeyish/smartdns-rs#203] [General]
为啥需要它啊？

**Comment by mokeyish** [comment]
你是 gnu 静态编译？换成 musl 吧，musl 可以在 openwrt 上运行的。

**Reply by PikuZheng** [reply]
就是musl做不到纯静态
大体上是这个命令`RUSTFLAGS="-C linker=gcc -C target-feature=+crt-static" cargo build --release --target=i686-unknown-linux-musl`（也不知道那个用户为啥需要x86架构的）

**Reply by mokeyish** [reply]
32 位的 x86? 这个我也没编译过

**Reply by PikuZheng** [reply]
大致看了下好像openwrt的启动脚本、linux的服务安装和ipt/nft检测用到 which，希望能把启动脚本、安装脚本从主程序中分离出来，毕竟容器运行这些都不需要

**Reply by PikuZheng** [reply]
所以现在实现了跨平台，但是没有实现跨架构 🤣

**Comment by mokeyish** [comment]
可以去掉的。增加 nft 和 service 两个 feature 把相关代码用 feature 来启用。今天写了一天，我改不动了，你看下能不能改吧。参考这个  legacy_dns_server

https://github.com/mokeyish/smartdns-rs/blob/main/src/server/mod.rs#L5

**Comment by mokeyish** [comment]
这里拆分了。 https://github.com/mokeyish/smartdns-rs/pull/207

**Reply by PikuZheng** [reply]
```
# file smartdns
smartdns: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), statically linked, stripped
```
成了，不过又发现了其他问题：
```
├── hickory-proto v0.24.0 (/workdir/target/patch/hickory-proto-0.24.0)
│   ├── h3-quinn v0.0.3
│   │   ├── quinn v0.10.2
│   │   │   ├── quinn-proto v0.10.6
│   │   │   │   ├── ring v0.16.20
```
这个ring 0.16 只支持x86(64)和arm(64)，以至于mips之类的没法用。另一方面quinn好像不支持tls1.3 quic，而且可能一时更新不了

**Reply by mokeyish** [reply]
ring 0.17 好像 windows arm64 也支持了。不过我暂时不急，先不搞。

**Reply by PikuZheng** [reply]
这个是quinn的锅，事实上现在quinn同时需要ring0.16和0.17，而且一时半会儿不能抛弃0.16（需要大改
